README.md

This awesome library provides an expressive way to specify, run and debug **finite state machines** in Javascript.

Here are the main features:

* Express compact state transitions with regular expressions!
* Debug your FSM on line with a mini-server (powered by `socketio`) with D3js rendering. 

You can see above a live demonstration of what you'll see when you connect to your fsm with a browser [^1]. And yes, **you can attach to different FSMs simultaneously** by using different ports.

**Note**: The fsm runs server-side, on node! This is not compatible with browsers at the moment.

## Installation

To install, use `npm`:

    npm install fluent-fsm


## Usage

Import the prototype in your program:

    fsm = require('fsmexpress').fsm;

### Create fsm and instantiate states

Create a finite state machine:

    fs = new fsm()

Define states (`livescript/coffeescript` code):

    fs.define-as-states([   'II' 'SI' 'PI' 'OI' 
                            'IS' 'SS' 'PS' 'OS' 
                            'IP' 'SP' 'PP' 'OP' 
                            'IC' 'SC' 'PC' 'OC' 'error' ])
                            
    fs.define-as-initial('II')

### Define transitions
Define a transition (optionally using a regular expression) from all states beginning with `I` excluding some states (`IP`, `IC`) on a specific event (`an_event`) and register function `action_to_trigger`  to be triggered contextually:

    fs.add_rule 
        from: 'I(.+)'
        excluding: 'IP IC' 
        at: 'an_event' 
        jumpTo: 'S-' 
        execute: action_to_trigger  
    

**Note**: the target state `S-` is a state beginning with `S` and ending with the matched text in `(.+)` in the `from` expression. So the above statement will generate only two different state transitions (because `'IP' 'IC'` are not allowed `from` states):

    II -> SI
    IS -> SS
    

### Unfold and optimize 

After the state transitions have been setup, invoke `unfold` to generate actual state transition rules:

    fs.unfold()

Prune states that are not reachable:
    
    fs.optimize()
    
    
### Linking to an event emitter

To register an event emitter:

    fs.registerEventEmitter(the_event_emitter)

`the_event_emitter` should be a Node `Emitter` object. The FSM registers its own listeners to enable state transition internal methods. Practically, let's assume that we have the following event emitter:

    class tester extends EventEmitter
        
        run_op: ~> 
            @emit 'anEvent'
            setTimeout(@run_tr, 300)
        
        run_tr: ~> 
            @emit 'anEvent2'
            setTimeout(@run_fl, 300)
        
        run_fl: ~> 
            @emit 'anotherEvent'
            setTimeout(@run_op, 300)

Let's register it and start the finite state machine:

    tst = new tester()
    
    # Register event emitter and start the fsm
    fs.registerEventEmitter(tst)  
    fs.start()   
    

## GUI debug

You can have a visual representation of the FSM that is served through a small web service (screenshot above):

    red = "#9d261d"
    gre = "#46a546"
    blu = "#049cdb"

    # GUI related stuff..
    fs.prepare-emit()
    fs.mark transition: '.+',       with-color: 'lightgrey'
    fs.mark transition: '.+Open',   with-color: "#gre"
    fs.mark transition: '.+Close',  with-dashed-color: "#gre"
    fs.mark transition: 'failed.+', with-color: "indianred"
    fs.mark state:      '.+',       with-color: 'lightgrey'
    fs.mark state:      'error',    with-color: 'indianred'
    fs.mark state:      fs.initial, with-color: "#gre"
    fs.mark state:      fs.final,   with-color: "lightsteelblue"
    console.log fs.data

    fs.serve(6970, 'my fsm')

You can see live state transitions (wherever the fsm is, even remotely, provided that the port can be accessed).

## License
MIT


 [^1]: Event generation is simulated in this page. 

history.txt
418 clone
597 clone
2227 show-branch
2234 branch
2235 co
2237 help
2238 branch
2242 ignore
2246 commit
2247 help
2251 tag
2252 commit
2253 whatchanged
2655 tag
2656 head
2681 commit
2682 push
2683 merge
2684 help
2760 help
2762 checkout
3071 push
3259 annex
3292 --shortstat
3293 log
3294 log-json
3295 log-json
3299 log-json
3300 log
3301 merge
3302 log-json
3303 log-json
3766 update
3814 tag
3821 diff
4374 checkout
4375 tags
4603 rm
4604 rm
4605 rm
4607 commit
4608 commit
4609 commit
4610 add
4611 add
4612 add
4618 checkout
4622 statsu
4623 merge
4625 push
4626 push
5657 clone
5736 status
5969 checkout
6751 heads
7112 clone
7317 pull
7318 checkout


src/example.js
(function(){
  var fsm, anyOf, EventEmitter, fs, red, gre, blu, tester, tst;
  fsm = require('../lib/fsmexpress.js').fsm;
  anyOf = require('../lib/fsmexpress.js').anyOf;
  EventEmitter = require('events').EventEmitter;
  fs = new fsm();
  fs.defineAsStates(['II', 'SI', 'PI', 'OI', 'IS', 'SS', 'PS', 'OS', 'IP', 'SP', 'PP', 'OP', 'IC', 'SC', 'PC', 'OC', 'error']);
  fs.defineAsInitial('II');
  fs.from('I(.+)').butNotFrom(anyOf(['IP', 'IC'])).on('anEvent').nextIs('S-').otherwiseIs('error');
  fs.from('S(.+)').butNotFrom(anyOf(['SP', 'SC'])).on('anEvent2').nextIs('P-').otherwiseIs('error');
  fs.from('P(.+)').butNotFrom(anyOf(['PC'])).on('anEvent3').nextIs('O-').otherwiseIs('error');
  fs.from('P(.+)').butNotFrom(anyOf(['PP', 'PC'])).on('anotherEvent').nextIs('II').otherwiseIs('error');
  fs.from('(.+)I').on('anotherEvent2').nextIs('-S').otherwiseIs('error');
  fs.from('(.+)S').butNotFrom(anyOf(['IS', 'SS', 'PS'])).on('anotherEvent3').nextIs('-P').otherwiseIs('error');
  fs.from('OP').on('differentEvent').nextIs('II');
  fs.from('(.+)P').on('differentEvent2').nextIs('-I');
  fs.unfold();
  fs.optimize();
  red = "#9d261d";
  gre = "#46a546";
  blu = "#049cdb";
  fs.prepareEmit();
  fs.mark({
    transition: '.+',
    withColor: 'lightgrey'
  });
  fs.mark({
    transition: '.+Open',
    withColor: gre + ""
  });
  fs.mark({
    transition: '.+Close',
    withDashedColor: gre + ""
  });
  fs.mark({
    transition: 'failed.+',
    withColor: "indianred"
  });
  fs.mark({
    state: '.+',
    withColor: 'lightgrey'
  });
  fs.mark({
    state: 'error',
    withColor: 'indianred'
  });
  fs.mark({
    state: fs.initial,
    withColor: gre + ""
  });
  fs.mark({
    state: fs.final,
    withColor: "lightsteelblue"
  });
  console.log(fs.data);
  fs.serve(6970, 'my fsm');
  tester = (function(superclass){
    var prototype = extend$((import$(tester, superclass).displayName = 'tester', tester), superclass).prototype, constructor = tester;
    prototype.run_op = function(){
      this.emit('anEvent');
      return setTimeout(this.run_tr, 300);
    };
    prototype.run_tr = function(){
      this.emit('anEvent2');
      return setTimeout(this.run_fl, 300);
    };
    prototype.run_fl = function(){
      this.emit('anotherEvent');
      return setTimeout(this.run_op, 300);
    };
    function tester(){
      this.run_fl = bind$(this, 'run_fl', prototype);
      this.run_tr = bind$(this, 'run_tr', prototype);
      this.run_op = bind$(this, 'run_op', prototype);
      tester.superclass.apply(this, arguments);
    }
    return tester;
  }(EventEmitter));
  tst = new tester();
  fs.registerEventEmitter(tst);
  fs.start();
  tst.run_op();
  function bind$(obj, key, target){
    return function(){ return (target || obj)[key].apply(obj, arguments) };
  }
  function extend$(sub, sup){
    function fun(){} fun.prototype = (sub.superclass = sup).prototype;
    (sub.prototype = new fun).constructor = sub;
    if (typeof sup.extended == 'function') sup.extended(sub);
    return sub;
  }
  function import$(obj, src){
    var own = {}.hasOwnProperty;
    for (var key in src) if (own.call(src, key)) obj[key] = src[key];
    return obj;
  }
}).call(this);
